File: java_solutions/5.longest-palindromic-substring.java
/*
 * @lc app=leetcode id=5 lang=java
 *
 * [5] Longest Palindromic Substring
 */

// @lc code=start
// semicolons: 1
class Solution {
    public String longestPalindrome(String s) {
        return IntStream.range(0, 2 * s.length()).mapToObj(i -> Stream.iterate(new int[]{i / 2, i / 2 + i % 2}, p -> p[0] >= 0 && p[1] < s.length() && s.charAt(p[0]) == s.charAt(p[1]), p -> new int[]{p[0] - 1, p[1] + 1}).reduce((a, b) -> b).orElse(new int[]{i / 2, i / 2 + i % 2 - 1})).max(Comparator.comparingInt(p -> p[1] - p[0])).map(p -> s.substring(p[0], p[1] + 1)).orElse("");
    }
}
// @lc code=end

File: kotlin_solutions/5.longest-palindromic-substring.kt
/*
 * @lc app=leetcode id=5 lang=kotlin
 *
 * [5] Longest Palindromic Substring
 */

// @lc code=start
class Solution { fun longestPalindrome(s: String): String = Array(s.length) { i -> BooleanArray(s.length) { i >= it } }.let { dp -> s.indices.fold(s.take(1)) { res, to -> (to - 1 downTo 0).fold(res) { r, from -> dp[from][to].let { dp[from][to] = s[from] == s[to] && dp[from + 1][to - 1] } .let { if (dp[from][to] && to - from + 1 > r.length) s.substring(from, to + 1) else r } } } } }
// @lc code=end

